### ISSUE TITLE
Tree shaking

### ISSUE THREAD

**Issue by raphaeltm** [run-llama/LlamaIndexTS#395]
Using LlamaIndex with Next.js deployed to Vercel resulted in loads of warnings like this one for me. Also, it more than quadrupled my deployment times from 3min to 13min:

```
Warning: Max serverless function size of 50 MB compressed or 250 MB uncompressed reached
--
15:46:26.417 | Serverless Function's page: _error.js
15:46:26.419 | Large Dependencies                 Uncompressed size  Compressed size
15:46:26.419 | node_modules/canvas/build                  178.98 MB         45.74 MB
15:46:26.419 | .next/server/chunks                          8.01 MB          2.61 MB
15:46:26.419 | node_modules/next/dist                     709.62 KB        215.13 KB
15:46:26.419 | node_modules/iconv-lite/encodings          309.99 KB        171.38 KB
15:46:26.420 | node_modules/react-dom/cjs                 544.24 KB        138.16 KB
15:46:26.420 |  
15:46:26.420 | All dependencies                           188.82 MB         48.99 MB
```

`canvas` is a dependency of the `pdfjs-dist` package which is a `llamaindex` dependency. It is massive.

From what I can tell, it seems like `llamaindex` doesn't support tree shaking, which means that even if you don't use any of the pdf parsing capabilities, you still end up with the massive dependencies, which could cause issues with platforms like Vercel. 

I got around it by adding `"postinstall": "rimraf node_modules/canvas",` to my package, but ideally the LlamaIndex package would just support tree shaking so that wouldn't be necessary.

**Comment by himself65** [comment]
Thanks, I'm working on this

See #391

**Comment by owencraston** [comment]
I am also experiencing the same issue. Looking forward to when this fix ships.

**Comment by owencraston** [comment]
@himself65 is it going to be possible to reduce the bundle size even if I need the PDF parsing capabilities?

**Comment by himself65** [comment]
First I will move to the peer dips, and second I might only reduce about 10MB if you really need it

**Comment by himself65** [comment]
Might related: https://github.com/vercel/next.js/issues/58313#issuecomment-1883861427

**Comment by owencraston** [comment]
As soon as I added the current llama index package to my repo I started getting this error in vercel.

`Error: A Serverless Function has exceeded the unzipped maximum size of 250 MB. : https://vercel.link/serverless-function-size`

If I remove the library the issue goes away.

**Comment by himself65** [comment]
Related: https://github.com/run-llama/LlamaIndexTS/issues/394

**Comment by himself65** [comment]
could you give me the reproduction code to test if this still happen?

**Comment by owencraston** [comment]
@himself65 I don't have the exact steps at this moment since my repo was private and deployed to vercel. I will try to get those for you soon.

If you are feeling adventurous this is roughly what I did....
1. create a [t3 stack](https://create.t3.gg/) app: you can use this one: https://github.com/owencraston/test-app
2. add "llamaindex": "0.0.44", to the package.json
3. add a few pdfs to the codebase
    - https://cpcassets.conservative.ca/wp-content/uploads/2023/11/23175001/990863517f7a575.pdf
    - https://2023.liberal.ca/wp-content/uploads/sites/565/2023/05/Policy-Resolutions-2023-National-Convention_OFFICIAL_ENG.pdf
    - https://www.greenparty.ca/sites/default/files/gpc_platform_en_v-02.pdf
4. deploy to vercel

**Comment by himself65** [comment]
Thank u I will try that

**Comment by himself65** [comment]
I'm working on this today, hope I can get things done

**Comment by himself65** [comment]
I just realized this is not just a PDF bundle issue; it's more like an API design issue in the server.

1. If you already have some pdf and want to do some RAG

Pre-compile (do embedding or something) before building the server, meaning the pdf parser won't be used in the server runtime. In this case, next.js will tree-shake the packages correctly because you didn't use it.

2. If you need the user to upload a PDF and do RAG base on that.

For this case, I realized that `PDFReader` is not good for this case since it can only accept a local file (by design) (see type below)

```ts
declare class PDFReader implements BaseReader {
    loadData(file: string, fs?: GenericFileSystem): Promise<Document[]>;
}
```

I'm considering accepting UInt8Array for input. But pdfjs-dist cannot be tree shaked. and this is Vercel's issue because they are using AWS serverless under the hood. (https://github.com/orgs/vercel/discussions/103)

Possible solutions: 

1. **use other web hosting services**, and create a real server instance.
2. **put pdfjs-dist into client side**, and send the render result to the server

But in real production, there should be a database/storage (like Cloudflare R2 or Postgres)

**Comment by himself65** [comment]
Fixed in #477

**Comment by himself65** [comment]
We use `pdf2json` now, and seems like it works well on the vercel server side; please try the new version `llamaindex@0.1.4`.

Report to us if you see the same issue
